### ISSUE TITLE
Refactor Model Class

### ISSUE THREAD

**Issue by sgill2** [MobleyLab/blues#74]
At some point soon we're looking to refactor the model class so that it can modularly combine different types of moves by using a redefined Class structure and generators (See openforcefield).

**Comment by sgill2** [comment]
@davidlmobley Could you link the generator example from the openforcefield github for reference when you have a chance to?

**Comment by davidlmobley** [comment]
@sgill2 https://github.com/open-forcefield-group/openforcefield/blob/master/openforcefield/typing/engines/smirnoff/forcefield.py -- search for generators. Basically towards the end of the module it's all definitions of generators.

**Comment by sgill2** [comment]
After thinking about the `Model` changes, this is the cleanest implementation I came up with, which would change the following:

1. **Move all the different move types into their own class which inherits off of `Model`** 

The base `Model` class itself would not innately have a method which performs the desired move, but would have other methods that would be generally helpful to have (such as the `getAtomIndices()` and `getMasses()`). The move itself would be performed by a method with the same name across all children classes called `move()`. Currently `MoveProposal` handles the position changes associated with the move.

2. **Allow multiple Models to be passed into `MoveProposal`**
Currently `MoveProposal` takes in only one `Model` class at a time. To allow the user to mix and match different types of moves I'd like to change it so `MoveProposal` can take in a set of `Model` instances specified by a list or tuple. This set can follow the following conventions:
If the `Model` are in a tuple then the move is performed in the order listed. If the `Model` is in a list then one `Model` out of the ones in the list is randomly chosen to be performed.

For example, say we have A, B, and C `Model` objects. If `(A,B,C)` is passed into `MoveProposal` then during the NCMC simulation `A.move()` will be performed followed by `B.move()` then `C.move()`. 
However if `[A,B,C]` is passed in, then A **OR** B **OR** C move method will be performed. 
It could also be possible to combine these two specifications, so if you passed – say, `[A, (B,C)]` – then during the move proposal step either `A.move()` **OR**  `B.move()` **AND** `C.move()` would be performed.

Any thoughts/input on this @nathanmlim @davidlmobley ?

**Comment by davidlmobley** [comment]
@sgill2 - I'm confused by this. Trying to talk to you IRL.

**Comment by sgill2** [comment]
My pseudocode to illustrate how I imagine the result looking:
(let A = `Model` for ligand rotation)
(let B = `Model` for side chain rotation)
```python
A = A(*args)
B = B(*args)
move_list = (A, B)
move_prop = MoveProposal(move_list, nstepsNC)
Simulation(move_prop, everything_else)
Simulation.run()
```

**Comment by nathanmlim** [comment]
In the simulation.py script [here](https://github.com/MobleyLab/blues/blob/master/blues/simulation.py#L244) we have the lines:

```python
self.model, self.nc_context = self.mover.moves['method'](model=self.model, nc_context=self.nc_context)
```

In `mover.moves` is a dictionary that stores the list of methods to apply onto the model object. If you want to apply multiple move types on a single ligand you could loop over them like:

```python
for m in mover.moves['method']:
    m(model,nc_context)
```

**Comment by davidlmobley** [comment]
@sgill2 - the thing your approach does is weds us to one particularly handling of a `move_list` in `MoveProposal`, which is less than ideal, I think. What if I want to spend 80% of my time doing ligand rotational moves, 10% of my time doing smart darting moves, and 10% of it doing sidechain rotations? Seems like this would be fairly hard to code up in this framework.

**Comment by sgill2** [comment]
You could always assign a probability attribute to each Model object that gets checked if you want to do different probability moves. Right now we don't have anything in place to handle that case anyway.

**Comment by davidlmobley** [comment]
I think what I'm trying to say is that I like the relatively clean behavior of Nathan's existing code. Specifically, @sgill2 , if you wanted to mix multiple kinds of moves in Nathan's existing framework I think you could do it with something like:

```python
for m in mover.moves['method']:
   self.model, self.nc_context = m(model, nc_context)
```
as Nathan wrote, which is super minimal. But if you wanted to be fancy about it, it would be easy to mix different probabilities:

```python
rigid_rotations = [ m for m in mover.moves if m.name=='rigid_rotation']
sidechain_rotations = [ m for m in mover.moves if m.name=='sidechain_rotation']
r = rand()
if r < 0.2:
  for m in sidechain_rotations:
      self.model, self.nc_context = m(model, nc_context)
else:
  for m in rigid_rotations:
      self.model, self.nc_context = m(model, nc_context)
```

That just seems clean to me. 

Unless there is a very strong argument for making the `move` be a function of the `Model`, @sgill2 , I'd really prefer to have it remain separate -- to me, conceptually, a move is something which modifies a model, as in `self.model, self.nc_context = m(model, nc_context)`, but it is not a property/function of a model itself. So it should logically be separate unless there is a profound reason for connecting it.

**Comment by sgill2** [comment]
Just to be clear let me start by reiterating the points I'm trying to make
1. I think that each type of move to be performed should have it's own separate model
2. I think that the move should be tied to the model

@davidlmobley I'm not seeing how it would be easier. If you were going to propose probabilities in that fashion you would have to change the *base code every time you want to change your probabilities** since that code appears in `Simulation`. Also I still don't think that suggestion handles multiple regions very gracefully. 
>rigid_rotations = [ m for m in mover.moves if m.name=='rigid_rotation']
sidechain_rotations = [ m for m in mover.moves if m.name=='sidechain_rotation']
r = rand()
if r < 0.2:
  for m in sidechain_rotations:
      self.model, self.nc_context = m(model, nc_context)
else:
  for m in rigid_rotations:
      self.model, self.nc_context = m(model, nc_context)

Say you wanted to select one sidechain to be rotated with a probability of 0.8 and another with a probability of 0.2; how would you specify that since you'd be giving all `sidechain_rotation` moves the same probability? Separate models with their own probabilities assigned wouldn't have that problem.

Specifying a model for each type of move also has another benefit for developers. You pointed out that it would be possible to specify multiple ligands in `rigid_rotation` – if the user had foresight to do that. Generalizing code can take quite a bit of work, and allowing separate models for each move makes testing ideas easier.

Also by tying in the move to `Model` it's much easier for a user to implement their own moves. If a user wants to make their own move in the current framework, first they have to make their own `Model` class with the proper attributes. THEN they also have to modify `MoveProposal` so that it can actually perform the move. 
It seems like an unnecessary hoop to go through making two custom classes – or changes to the master branch – for one desired modification. At the very least that was my experience while refactoring the smart darting move.  I think lowering barriers to usable code is a good thing.

I feel like I'm more adamant about allowing multiple models to be specified in the move proposal. It mostly just irks me I have to make changes in multiple places when I want to implement a move that requires a new model. That, and it makes more conceptual sense to me that one `Model` = one move.

**Comment by davidlmobley** [comment]
I agree that my way wouldn't make it super easy to combine multiple moves with differing probabilities without modifying `Simulation`. I'm just pointing out that how you'd modify `Simulation` would be straightforward. 

However, @sgill2 - I think you're thinking of the world we're in now, where we don't have many `Model` classes implemented so each new kind of move involves a new `Model` AND a new way to propose moves, whereas I'm thinking of the world we're going to be in soon where we want to implement new kinds of move proposals that use properties that are already held by `Model`s we already have. For example, we're going to be doing sidechain rotational moves where we propose random sidechain rotations. But we'll also want to try generating the rotational moves in other ways -- for example, the move proposal engine can match the sidechain being moved to a rotamer library and use that to bias the angles proposed for the move. This would use the same `Model` and only the move proposal would be different. 

More generally, I think we'll want to ultimately be exploring different ways of making the same general kind of move proposals which would use the same input `Model` but do something different with it. This seems to me to be a strong argument for having `Model` and move proposals be separate. 

> Also by tying in the move to Model it's much easier for a user to implement their own moves. If a user wants to make their own move in the current framework, first they have to make their own Model class with the proper attributes. THEN they also have to modify MoveProposal so that it can actually perform the move.

This is not "easier to implement" in terms of "substantially less code", I don't think, I think it's just "fewer places to modify the code".  Or am I missing something? 

> It seems like an unnecessary hoop to go through making two custom classes – or changes to the master branch – for one desired modification. At the very least that was my experience while refactoring the smart darting move. I think lowering barriers to usable code is a good thing.

Yes, but I'm fairly convinced that keeping these things separate is going to help extensibility -- as argued above, think of the case where we already have a `Model` which contains the info you need so you only have to implement the requisite move, which is easier than writing a new `Model` AND move. And if each new type of move needs a new type of `Model` even if the `Model` is the same, we're going to accumulate a lot of code duplication, I think. 

Finally, I still think these are logically separate things so they should be constructed that way. A move is something which changes the system. A `Model` has properties of the system. 

I'm still open to counterarguments, and we should keep going until we're convinced one way or the other. :) This is an important decision for the long-term future of the code.

**Comment by sgill2** [comment]
Closed via #76
